### ISSUE TITLE
Can'r run prune and quantization together

### ISSUE THREAD

**Issue by lijian18** [NervanaSystems/distiller#34]
Hi,

When I try to run prune and quantization together on resnet20_cifar in one yaml, it failed. It said Key Error on xxx_float_weight. So what's the correct procedure to mix both of them together?

Thomas

**Comment by lijian18** [comment]
Traceback (most recent call last):
  File "compress_classifier.py", line 520, in <module>
    main()
  File "compress_classifier.py", line 321, in main
    compression_scheduler.on_epoch_begin(epoch)
  File "/home/thbian/distiller/distiller/scheduler.py", line 98, in on_epoch_begin
    policy.on_epoch_begin(self.model, self.zeros_mask_dict, meta)
  File "/home/thbian/distiller/distiller/policy.py", line 82, in on_epoch_begin
    self.pruner.set_param_mask(param, param_name, zeros_mask_dict, meta)
  File "/home/thbian/distiller/distiller/pruning/sensitivity_pruner.py", line 61, in set_param_mask
    zeros_mask_dict[param_name].mask = distiller.threshold_mask(param.data, threshold)
KeyError: 'module.layer3.1.conv1.float_weight'

My Yaml is an experimental one to see if it works, listed below.

pruners:
  pruner1:
    class: 'SensitivityPruner'
    sensitivities:
      'module.layer3.1.conv1.weight'      : 0.975
      'module.layer3.1.conv1.float_weight': 0.975

quantizers:
  dorefa_quantizer:
    class: DorefaQuantizer
    # class: SymmetricLinearQuantizer
    bits_activations: 8
    bits_weights: 3
    bits_overrides:
    # Don't quantize first and last layer
      conv1:
        wts: null
        acts: null
      layer1.0.pre_relu:
        wts: null
        acts: null
      final_relu:
        wts: null
        acts: null
      fc:
        wts: null
        acts: null

lr_schedulers:
  training_lr:
    class: MultiStepMultiGammaLR
    milestones: [80, 120, 160]
    gammas: [0.1, 0.1, 0.2]

policies:
  - pruner:
      instance_name : 'pruner1'
    starting_epoch: 2
    ending_epoch: 6
    frequency: 1

  - quantizer:
      instance_name: dorefa_quantizer
    starting_epoch: 0
    ending_epoch: 200
    frequency: 1

  - lr_scheduler:
      instance_name: training_lr
    starting_epoch: 0
    ending_epoch: 161
    frequency: 1

**Comment by guyjacob** [comment]
Hi Thomas,

Combining pruning and quantization is most definitely something we intend to explore, however it'll take some time before we're able to put meaningful focus on it.
Having said that, I'll try to provide some direction based on your question, and I'll keep the issue open if you or anyone wants to continue experimenting with this. But more bumps along the road are to be expected.

The reason for the exception you're getting is that the model is being modified for quantization only after the mapping between parameter names to the "zero masks" in the compression scheduler is created. So, the new `float_weight` parameters created by the quantizer don't exist in the mapping.
This can be seen in the `dict_config()` function `distiller/config.py`. The `CompressionScheduler` instance named `scheduler` is created right at the beginning, and that's when the mapping is created. The `QuantizationPolicy` is created later, and that's when the model modification takes place. 
So, you can try changing the order, which should at least get you past the exception.

Also note that it's probably not correct to specify both `float_weight` and `weight` of the same layer for pruning, as you do in your YAML. They are dependent on each other, so there's really no point in pruning each of them separately. Furthermore, since their distributions will likely not be the same, pruning each separately might results in inconsistencies in terms of which elements are pruned in each. See [here](https://nervanasystems.github.io/distiller/quantization/index.html#training-with-quantization) for more details on the connection between the float and quantized copies of the weights.

**Comment by lijian18** [comment]
Hi Jacob,

It's nice to see your helpful suggestion. I also thought leave float_weight there is unreasonable.

The reason why I did quantization and prune together is, normally we want to do prune first and then quantization, retrain, and then iteratively continue that loop.

What I saw is, 1) Quantization must start from epoch 0; 2) It can't mix with prune; 3) There's no weight dump procedure to load/save network weights. Continue training from weights instead of checkpoint is also workable, right? So, is there any plan to improve all of them or I am looking at the wrong directions?

My confusion is, if I need to do quantization first, and then prune, the quantization will be messed by prune. But I can't quantize it again because quantization must start from epoch 0. Is there a solution to that? 

Regards,
Thomas 


------------------------------------------------------------------
发件人：Guy Jacob <notifications@github.com>
发送时间：2018年7月25日(星期三) 06:58
收件人：NervanaSystems/distiller <distiller@noreply.github.com>
抄　送：lijian18 <lijian_18@aliyun.com>; Author <author@noreply.github.com>
主　题：Re: [NervanaSystems/distiller] Can'r run prune and quantization together (#34)

Hi Thomas,
Combining pruning and quantization is most definitely something we intend to explore, however it'll take some time before we're able to put meaningful focus on it.
 Having said that, I'll try to provide some direction based on your question, and I'll keep the issue open if you or anyone wants to continue experimenting with this. But more bumps along the road are to be expected.
The reason for the exception you're getting is that the model is being modified for quantization only after the mapping between parameter names to the "zero masks" in the compression scheduler is created. So, the new float_weight parameters created by the quantizer don't exist in the mapping.
 This can be seen in the dict_config() function distiller/config.py. The CompressionScheduler instance named scheduler is created right at the beginning, and that's when the mapping is created. The QuantizationPolicy is created later, and that's when the model modification takes place.
 So, you can try changing the order, which should at least get you past the exception.
Also note that it's probably not correct to specify both float_weight and weight of the same layer for pruning, as you do in your YAML. They are dependent on each other, so there's really no point in pruning each of them separately. Furthermore, since their distributions will likely not be the same, pruning each separately might results in inconsistencies in terms of which elements are pruned in each. See here for more details on the connection between the float and quantized copies of the weights.

**Comment by guyjacob** [comment]
Actually, we did a little bit of refactoring recently in terms of how the changes to the Optimizer are handled which, I think, could allow us to start quantization in epoch > 0. But I'll need to check.

Also - if you want to prune and then quantize, then you should be able to do that in right now with separate runs. Do a run with just pruning, then do a run with quantization and use '--resume' to load the pruned weights as a starting point. That should work.

Resuming from the quantization run is a problem, see my comment [here](https://github.com/NervanaSystems/distiller/issues/21#issuecomment-403293755). Perhaps a workaround you can try for now, is to modify the saved checkpoint before resuming from it:

1. For each pair of "float_weight" and "weight", copy the contents "float_weight" into "weight". Then you can either discard the "float_weight" entry, or alternatively add `strict=False` to the call to `load_state_dict` at https://github.com/NervanaSystems/distiller/blob/master/apputils/checkpoint.py#L121
2. Remove the `quantizer_metadata` entry from the checkpoint, or alternatively comment out the code that loads it at https://github.com/NervanaSystems/distiller/blob/master/apputils/checkpoint.py#L113

These 2 steps above should make it seem as if you're resuming from a "regular" FP32 model. Define the quantization policy in the YAML as you'd normally do.

Regarding your comment on "no weight dump procedure" - not sure I understand what you mean. What's wrong with the checkpoints as they are now? That's the standard PyTorch mechanism for saving/loading weights.

Thanks,
Guy

**Comment by lijian18** [comment]
Hi,

Actually I did a little modification to some of source files and now I can run quantization and prune together, I verified the results and it worked. I still start quantization from epoch 0 though because weight replacing process better be kicked off from the beginning. I found that weights are all removed from model so prune must work on float_weight instead, below is what I modified.

1) Add one more function to class CompressionScheduler; 

    def append_float_weight_after_quantizer (self):
        for name, param in self.model.named_parameters():
            if name in self.zeros_mask_dict:
                continue
            masker = ParameterMasker(name)
            self.zeros_mask_dict[name] = masker
            param.register_hook(partial(masker.apply_mask, in_backward_cb=True))
        return

2) Call the function in compress_classifier.py after quantizer is created (this is why it has to be epoch zero) in order to add float_weight into zeros_mask_dict.

    if args.compress:
        # The main use-case for this sample application is CNN compression. Compression
        # requires a compression schedule configuration file in YAML.
        compression_scheduler = distiller.file_config(model, optimizer, args.compress)
        compression_scheduler.append_float_weight_after_quantizer ()

3) Change weights to float_weights counterpart in yaml file.

      e.g. 'module.layer1.0.conv1.weight': 0.85 -> 'module.layer1.0.conv1.float_weight': 0.85

4) In log we can see only float_weight is pruned. Actually adding weight has no effect.

Parameters:
+----+-------------------------------------------+----------------+---------------+----------------+------------+------------+----------+----------+----------+------------+---------+----------+------------+
|    | Name                                      | Shape          |   NNZ (dense) |   NNZ (sparse) |   Cols (%) |   Rows (%) |   Ch (%) |   2D (%) |   3D (%) |   Fine (%) |     Std |     Mean |   Abs-Mean |
|----+-------------------------------------------+----------------+---------------+----------------+------------+------------+----------+----------+----------+------------+---------+----------+------------|
|  0 | module.conv1.float_weight                 | (16, 3, 3, 3)  |           432 |            432 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 2.41392 |  0.50737 |    1.39273 |
|  1 | module.conv1.weight                       | (16, 3, 3, 3)  |           432 |            432 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.68345 |  0.07469 |    0.61019 |
|  2 | module.layer1.0.conv1.float_weight        | (16, 16, 3, 3) |          2304 |           1447 |    0.00000 |    0.00000 |  0.00000 |  0.78125 |  0.00000 |   37.19618 | 0.36077 | -0.05069 |    0.26119 |
|  3 | module.layer1.0.conv1.weight              | (16, 16, 3, 3) |          2304 |           2304 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.36434 | -0.02350 |    0.29149 |
|  4 | module.layer1.0.conv2.float_weight        | (16, 16, 3, 3) |          2304 |           1432 |    0.00000 |    0.00000 |  0.00000 |  1.17188 |  0.00000 |   37.84722 | 0.25623 | -0.00974 |    0.18413 |
|  5 | module.layer1.0.conv2.weight              | (16, 16, 3, 3) |          2304 |           2304 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.38430 |  0.01100 |    0.30197 |
|  6 | module.layer1.1.conv1.float_weight        | (16, 16, 3, 3) |          2304 |           1385 |    0.00000 |    0.00000 |  0.00000 |  1.17188 |  0.00000 |   39.88715 | 0.19228 | -0.01896 |    0.13566 |
|  7 | module.layer1.1.conv1.weight              | (16, 16, 3, 3) |          2304 |           2304 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.29413 | -0.00243 |    0.23304 |
|  8 | module.layer1.1.conv2.float_weight        | (16, 16, 3, 3) |          2304 |           1390 |    0.00000 |    0.00000 |  0.00000 |  0.78125 |  0.00000 |   39.67014 | 0.15979 | -0.00186 |    0.11251 |
|  9 | module.layer1.1.conv2.weight              | (16, 16, 3, 3) |          2304 |           2304 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.30217 |  0.02321 |    0.23854 |
| 10 | module.layer1.2.conv1.float_weight        | (16, 16, 3, 3) |          2304 |           1210 |    0.00000 |    0.00000 |  0.00000 |  1.56250 |  0.00000 |   47.48264 | 0.14345 | -0.00977 |    0.08817 |
| 11 | module.layer1.2.conv1.weight              | (16, 16, 3, 3) |          2304 |           2304 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.21276 |  0.01707 |    0.16111 |
| 12 | module.layer1.2.conv2.float_weight        | (16, 16, 3, 3) |          2304 |           1138 |    0.00000 |    0.00000 |  0.00000 |  1.95312 |  0.00000 |   50.60764 | 0.12927 | -0.00001 |    0.07749 |
| 13 | module.layer1.2.conv2.weight              | (16, 16, 3, 3) |          2304 |           2304 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.23198 |  0.03362 |    0.17269 |
| 14 | module.layer2.0.conv1.float_weight        | (32, 16, 3, 3) |          4608 |           1001 |    0.00000 |    0.00000 |  0.00000 | 34.17969 |  0.00000 |   78.27691 | 0.10546 | -0.00129 |    0.04648 |
| 15 | module.layer2.0.conv1.weight              | (32, 16, 3, 3) |          4608 |           4608 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.28256 |  0.04852 |    0.17590 |
| 16 | module.layer2.0.conv2.float_weight        | (32, 32, 3, 3) |          9216 |           2270 |    0.00000 |    0.00000 |  0.00000 | 12.79297 |  0.00000 |   75.36892 | 0.09280 | -0.00758 |    0.04421 |
| 17 | module.layer2.0.conv2.weight              | (32, 32, 3, 3) |          9216 |           9216 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.31366 |  0.02507 |    0.19685 |
| 18 | module.layer2.0.downsample.0.float_weight | (32, 16, 1, 1) |           512 |            512 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.23302 | -0.02065 |    0.17224 |
| 19 | module.layer2.0.downsample.0.weight       | (32, 16, 1, 1) |           512 |            512 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.35737 | -0.03281 |    0.27422 |
| 20 | module.layer2.1.conv1.float_weight        | (32, 32, 3, 3) |          9216 |           2306 |    0.00000 |    0.00000 |  0.00000 | 15.62500 |  0.00000 |   74.97830 | 0.09524 | -0.00623 |    0.04614 |
| 21 | module.layer2.1.conv1.weight              | (32, 32, 3, 3) |          9216 |           9216 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.34613 |  0.02785 |    0.21573 |
| 22 | module.layer2.1.conv2.float_weight        | (32, 32, 3, 3) |          9216 |           2310 |    0.00000 |    0.00000 |  0.00000 | 16.11328 |  0.00000 |   74.93490 | 0.09085 | -0.00968 |    0.04439 |
| 23 | module.layer2.1.conv2.weight              | (32, 32, 3, 3) |          9216 |           9216 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.35291 |  0.01301 |    0.21918 |
| 24 | module.layer2.2.conv1.float_weight        | (32, 32, 3, 3) |          9216 |           2279 |    0.00000 |    0.00000 |  0.00000 | 16.40625 |  0.00000 |   75.27127 | 0.11174 | -0.00737 |    0.05399 |
| 25 | module.layer2.2.conv1.weight              | (32, 32, 3, 3) |          9216 |           9216 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.35936 |  0.02678 |    0.22169 |
| 26 | module.layer2.2.conv2.float_weight        | (32, 32, 3, 3) |          9216 |           2277 |    0.00000 |    0.00000 |  0.00000 | 19.33594 |  0.00000 |   75.29297 | 0.09676 | -0.00817 |    0.04679 |
| 27 | module.layer2.2.conv2.weight              | (32, 32, 3, 3) |          9216 |           9216 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.35325 |  0.02095 |    0.21842 |
| 28 | module.layer3.0.conv1.float_weight        | (64, 32, 3, 3) |         18432 |           2888 |    0.00000 |    0.00000 |  0.00000 | 38.03711 |  0.00000 |   84.33160 | 0.07456 | -0.00530 |    0.02926 |
| 29 | module.layer3.0.conv1.weight              | (64, 32, 3, 3) |         18432 |          18432 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.33259 |  0.03270 |    0.18467 |
| 30 | module.layer3.0.conv2.float_weight        | (64, 64, 3, 3) |         36864 |           5711 |    0.00000 |    0.00000 |  0.00000 | 27.92969 |  0.00000 |   84.50792 | 0.07087 | -0.00980 |    0.02784 |
| 31 | module.layer3.0.conv2.weight              | (64, 64, 3, 3) |         36864 |          36864 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.32679 |  0.01228 |    0.18112 |
| 32 | module.layer3.0.downsample.0.float_weight | (64, 32, 1, 1) |          2048 |           2048 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.12919 | -0.00373 |    0.10517 |
| 33 | module.layer3.0.downsample.0.weight       | (64, 32, 1, 1) |          2048 |           2048 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.43578 | -0.01237 |    0.35710 |
| 34 | module.layer3.1.conv1.float_weight        | (64, 64, 3, 3) |         36864 |           5589 |    0.00000 |    0.00000 |  0.00000 | 29.27246 |  0.00000 |   84.83887 | 0.07411 | -0.00829 |    0.02871 |
| 35 | module.layer3.1.conv1.weight              | (64, 64, 3, 3) |         36864 |          36864 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.32853 |  0.02045 |    0.18084 |
| 36 | module.layer3.1.conv2.float_weight        | (64, 64, 3, 3) |         36864 |           5577 |    0.00000 |    0.00000 |  0.00000 | 31.81152 |  0.00000 |   84.87142 | 0.06996 | -0.00905 |    0.02714 |
| 37 | module.layer3.1.conv2.weight              | (64, 64, 3, 3) |         36864 |          36864 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.33182 |  0.01458 |    0.18200 |
| 38 | module.layer3.2.conv1.float_weight        | (64, 64, 3, 3) |         36864 |           5449 |    0.00000 |    0.00000 |  0.00000 | 39.40430 |  0.00000 |   85.21864 | 0.06295 | -0.00650 |    0.02405 |
| 39 | module.layer3.2.conv1.weight              | (64, 64, 3, 3) |         36864 |          36864 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.32690 |  0.02374 |    0.17903 |
| 40 | module.layer3.2.conv2.float_weight        | (64, 64, 3, 3) |         36864 |           5486 |    0.00000 |    0.00000 |  0.00000 | 51.31836 |  0.00000 |   85.11827 | 0.04261 | -0.00407 |    0.01630 |
| 41 | module.layer3.2.conv2.weight              | (64, 64, 3, 3) |         36864 |          36864 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.32763 |  0.02618 |    0.17957 |
| 42 | module.fc.float_weight                    | (10, 64)       |           640 |            640 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.73749 | -0.06864 |    0.65390 |
| 43 | module.fc.weight                          | (10, 64)       |           640 |            640 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |    0.00000 | 0.57305 | -0.06703 |    0.54570 |
| 44 | Total sparsity:                           | -              |        541792 |         325673 |    0.00000 |    0.00000 |  0.00000 |  0.00000 |  0.00000 |   39.88966 | 0.00000 |  0.00000 |    0.00000 |
+----+-------------------------------------------+----------------+---------------+----------------+------------+------------+----------+----------+----------+------------+---------+----------+------------+
2018-07-27 10:11:27,603 - Total sparsity: 39.89

For quantization, what I meant is since we have done a good job on it, shall we write a summary on it and double check how well are all these weights quantized? Currently quantization is quite in log, only pruner speaks out in log... 

Regards,
Thomas
------------------------------------------------------------------
发件人：Guy Jacob <notifications@github.com>
发送时间：2018年7月26日(星期四) 19:17
收件人：NervanaSystems/distiller <distiller@noreply.github.com>
抄　送：lijian18 <lijian_18@aliyun.com>; Author <author@noreply.github.com>
主　题：Re: [NervanaSystems/distiller] Can'r run prune and quantization together (#34)

Actually, we did a little bit of refactoring recently in terms of how the changes to the Optimizer are handled which, I think, could allow us to start quantization in epoch > 0. But I'll need to check.
Also - if you want to prune and then quantize, then you should be able to do that in right now with separate runs. Do a run with just pruning, then do a run with quantization and use '--resume' to load the pruned weights as a starting point. That should work.
Resuming from the quantization run is a problem, see my comment here. Perhaps a workaround you can try for now, is to modify the saved checkpoint before resuming from it:
For each pair of "float_weight" and "weight", copy the contents "float_weight" into "weight". Then you can either discard the "float_weight" entry, or alternatively add strict=False to the call to load_state_dict at https://github.com/NervanaSystems/distiller/blob/master/apputils/checkpoint.py#L121
Remove the quantizer_metadata entry from the checkpoint, or alternatively comment out the code that loads it at https://github.com/NervanaSystems/distiller/blob/master/apputils/checkpoint.py#L113
These 2 steps above should make it seem as if you're resuming from a "regular" FP32 model. Define the quantization policy in the YAML as you'd normally do.
Regarding your comment on "no weight dump procedure" - not sure I understand what you mean. What's wrong with the checkpoints as they are now? That's the standard PyTorch mechanism for saving/loading weights.
Thanks,
 Guy

**Comment by guyjacob** [comment]
Hi Thomas,

As a workaround for adding masks for the float parameters, what you describe in 1 and 2 looks functionally correct to me. And referencing only the float weights as you describe in 3 is also a step in the right direction. 
But - there's more to it. In some cases, depending on the method used, the zero masks are also used to zero the gradients of the pruned elements during backprop, effectively removing those elements from the learning process. Since the the float + quantized pairs represent the same parameter, they should share the mask. So the pruner will modify the mask for the float copy (as you specified in the YAML), but then you need to reflect that modification in the mask of the quantized copy, because that's the copy back-prop is executed on. That should be fairly easy to do as another step after the calls to 'apply_mask' in the policies.

Regarding the fact that you don't see any sparsity in the quantized weights - I assume you're using DoReFa for quantization? The way weights are quantized in that scheme (you can see the details [here](https://nervanasystems.github.io/distiller/algo_quantization/index.html#dorefa)), is that they are first quantized to the [0, 1] range - let's call that value `w_q_pre`. The final quantized value is `w_q = 2 * w_q_pre - 1`. Note that `w_q_pre` is one of `{0/(2^k - 1), 1/(2^k - 1), ... , 1}`. So while `w_q_pre` can get a value of 0, `w_q` actually can't. That's why you don't see zeros in the quantized copy.

As for your last comment about a summary for quantized weights, it's still not clear to me what you're looking for. In pruning we're indeed zero-ing elements based on some criteria, and the results vary. So it makes sense to check "how well we pruned". With quantization, on the other hand, all weights are quantized. So I'm not sure I understand what you mean by "how well are all these weights quantized".

**Comment by lijian18** [comment]
Hi,

Sorry, my typo, I typed too fast without double check. I meant "how well are all these weights quantized?", I think it's better reflected in the log. BTW I did use DoReFa for quantization.

And I'll make more changes to the code and give it another try according to your suggestion. Thank you. :-)

Regards,
Thomas


------------------------------------------------------------------
发件人：Guy Jacob <notifications@github.com>
发送时间：2018年7月29日(星期日) 16:45
收件人：NervanaSystems/distiller <distiller@noreply.github.com>
抄　送：lijian18 <lijian_18@aliyun.com>; Author <author@noreply.github.com>
主　题：Re: [NervanaSystems/distiller] Can'r run prune and quantization together (#34)

Hi Thomas,
As a workaround for adding masks for the float parameters, what you describe in 1 and 2 looks functionally correct to me. And referencing only the float weights as you describe in 3 is also a step in the right direction.
 But - there's more to it. In some cases, depending on the method used, the zero masks are also used to zero the gradients of the pruned elements during backprop, effectively removing those elements from the learning process. Since the the float + quantized pairs represent the same parameter, they should share the mask. So the pruner will modify the mask for the float copy (as you specified in the YAML), but then you need to reflect that modification in the mask of the quantized copy, because that's the copy back-prop is executed on. That should be fairly easy to do as another step after the calls to 'apply_mask' in the policies.
Regarding the fact that you don't see any sparsity in the quantized weights - I assume you're using DoReFa for quantization? The way weights are quantized in that scheme (you can see the details here), is that they are first quantized to the [0, 1] range - let's call that value w_q_pre. The final quantized value is w_q = 2 * w_q_pre - 1. Note that w_q_pre is one of {0/(2^k - 1), 1/(2^k - 1), ... , 1}. So while w_q_pre can get a value of 0, w_q actually can't. That's why you don't see zeros in the quantized copy.
As for your last comment about a summary for quantized weights, it's still not clear to me what you're looking for. In pruning we're indeed zero-ing elements based on some criteria, and the results vary. So it makes sense to check "how well we pruned". With quantization, on the other hand, all weights are quantized. So I'm not sure I understand what you mean by "how well are all these weights quantized".

**Comment by guyjacob** [comment]
Closing this for now, feel free to re-open (or start a new issue) if you have more questions

**Comment by dabadee** [comment]
Are there any updates on this ? Can we now prune, quantized networks ?
